UnbufferedSerial does not expose enable_* functions from SerialBase; behavior not predictable when exposed

<!--

   ************************************** WARNING **************************************

   The ciarcom bot parses this header automatically. Any deviation from the 
   template may cause the bot to automatically correct this header or may result in a 
   warning message, requesting updates.

   Please ensure all sections of the template below are filled in and no changes 
   are made to the template format. Only bugs should be raised here as issues. 
   Questions or enhancements should instead be raised on our forums:
   https://forums.mbed.com/ .

   *************************************************************************************

-->

### Description of defect

The documentation claims `UnbufferedSerial` has two functions, `enable_input` and `enable_output`. However, they are not defined in the base class and SerialBase is not accessible as it is a private base class. There is also an issue where using the enable_* functions on an STM board has no problem, but running on a Cypress board corrupts the port.

#### Target(s) affected by this defect ?

I tested on a CY8CPROTO_062_4343W and NUCLEO_F429ZI, but I assume all since it is a header issue.

#### Toolchain(s) (name and version) displaying this defect ?

GCC_ARM

#### What version of Mbed-os are you using (tag or sha) ?

6.5.0

#### What version(s) of tools are you using. List all that apply (E.g. mbed-cli)

mbed: 1.10.4

#### How is this defect reproduced ? 

~~~
#include "mbed.h"

UnbufferedSerial pc(USBTX, USBRX);

int
main()
{
	printf("Serial enabled, disabling.\n");
	pc.sync();
	pc.enable_input(false);
	pc.enable_output(false);
	pc.enable_input(true);
	pc.enable_output(true);
	printf("Serial enabled, this should be visible.\n");

}
~~~

This will fail to compile with this error

~~~
./main.cpp:13:5: error: request for member 'enable_output' is ambiguous
  pc.enable_output(true);
     ^~~~~~~~~~~~~
In file included from ./mbed-os/connectivity/netsocket/include/netsocket/nsapi_ppp.h:22,
                 from ./mbed-os/mbed.h:28,
                 from ./main.cpp:1:
./mbed-os/platform/include/platform/FileHandle.h:210:17: note: candidates are: 'virtual int mbed::FileHandle::enable_output(bool)'
     virtual int enable_output(bool enabled)
                 ^~~~~~~~~~~~~
In file included from ./mbed-os/drivers/include/drivers/UnbufferedSerial.h:26,
                 from ./mbed-os/mbed.h:71,
                 from ./main.cpp:1:
./mbed-os/drivers/include/drivers/SerialBase.h:144:10: note:                 'void mbed::SerialBase::enable_output(bool)'
     void enable_output(bool enable = true);
          ^~~~~~~~~~~~~
~~~

It can be fixed by adding these two lines to `UnbufferedSerial.h`:

~~~
    using SerialBase::enable_input;
    using SerialBase::enable_output;
~~~

Furthermore, this code works with the fixes on the NUCLEO, but it fails on the CYPRESS board (screen fills with garbage characters). I suspect the deinit/reinit of the serial object is the culprit in the Cypress case.

--- a/drivers/include/drivers/UnbufferedSerial.h
+++ b/drivers/include/drivers/UnbufferedSerial.h
@@ -155,14 +155,16 @@ public:
      */
     short poll(short events) const override;
 
-    using SerialBase::readable;
-    using SerialBase::writeable;
-    using SerialBase::format;
     using SerialBase::attach;
     using SerialBase::baud;
+    using SerialBase::enable_input;
+    using SerialBase::enable_output;
+    using SerialBase::format;
+    using SerialBase::readable;
+    using SerialBase::writeable;
+    using SerialBase::IrqCnt;
     using SerialBase::RxIrq;
     using SerialBase::TxIrq;
-    using SerialBase::IrqCnt;
 
 #if DEVICE_SERIAL_FC
     // For now use the base enum - but in future we may have extra options

--- a/drivers/include/drivers/UnbufferedSerial.h
+++ b/drivers/include/drivers/UnbufferedSerial.h
@@ -143,6 +143,35 @@ public:
         return 0;
     }
 
+    /** Enable or disable input
+     *
+     * Control enabling of device for input. This is primarily intended
+     * for temporary power-saving; the overall ability of the device to operate
+     * for input and/or output may be fixed at creation time, but this call can
+     * allow input to be temporarily disabled to permit power saving without
+     * losing device state.
+     *
+     *  @param enabled      true to enable input, false to disable.
+     *
+     *  @return             0 on success
+     *  @return             Negative error code on failure
+     */
+    int enable_input(bool enabled) override;
+
+    /** Enable or disable output
+     *
+     * Control enabling of device for output. This is primarily intended
+     * for temporary power-saving; the overall ability of the device to operate
+     * for input and/or output may be fixed at creation time, but this call can
+     * allow output to be temporarily disabled to permit power saving without
+     * losing device state.
+     *
+     *  @param enabled      true to enable output, false to disable.
+     *
+     *  @return             0 on success
+     *  @return             Negative error code on failure
+     */
+    int enable_output(bool enabled) override;
 
     /** Check for poll event flags
      * Check the events listed in events to see if data can be read or written
@@ -157,8 +186,6 @@ public:
 
     using SerialBase::attach;
     using SerialBase::baud;
-    using SerialBase::enable_input;
-    using SerialBase::enable_output;
     using SerialBase::format;
     using SerialBase::readable;
     using SerialBase::writeable;

--- a/drivers/source/UnbufferedSerial.cpp
+++ b/drivers/source/UnbufferedSerial.cpp
@@ -99,6 +99,24 @@ short UnbufferedSerial::poll(short events) const
     return revents;
 }
 
+int UnbufferedSerial::enable_input(bool enabled)
+{
+    lock();
+    SerialBase::enable_input(enabled);
+    unlock();
+
+    return 0;
+}
+
+int UnbufferedSerial::enable_output(bool enabled)
+{
+    lock();
+    SerialBase::enable_output(enabled);
+    unlock();
+
+    return 0;
+}
+
 #if DEVICE_SERIAL_FC
 void UnbufferedSerial::set_flow_control(Flow type, PinName flow1, PinName flow2)
 {

--- a/drivers/source/UnbufferedSerial.cpp
+++ b/drivers/source/UnbufferedSerial.cpp
@@ -101,18 +101,14 @@ short UnbufferedSerial::poll(short events) const
 
 int UnbufferedSerial::enable_input(bool enabled)
 {
-    lock();
     SerialBase::enable_input(enabled);
-    unlock();
 
     return 0;
 }
 
 int UnbufferedSerial::enable_output(bool enabled)
 {
-    lock();
     SerialBase::enable_output(enabled);
-    unlock();
 
     return 0;
 }

